Added detection for bun/node inside typescript SDK runtime

benjaminstrasser opened the pull request:
As discussed by #6591

### Changes
Tries to detect bun/node runtime through the following rules. If a detected runtime is not node or bun an error is raised.

If a package.json is present and the `dagger.runtime` field is set, the configured runtime is used
If a package-lock.json file is detected node is used a the runtime
if a bun.lockb file is detect bun is used as the runtime
if no runtime can be detected node is used by default

TomChv commented:
Hey! @benjaminstrasser 

I'm glad you opened a PR, I'll have a look at it today 🚀

TomChv commented on core/integration/module_typescript_test.go:
```diff
@@ -473,6 +473,128 @@ func TestModuleTypescriptOptional(t *testing.T) {
+	t.Run("should use package.json configuration node", func(t *testing.T) {
+		modGen := modGen.WithNewFile("/work/dagger/package.json", dagger.ContainerWithNewFileOpts{
+			Contents: `{
+				"name": "my-module",
+				"version": "1.0.0",
+				"description": "My module",
+				"main": "index.js",
+				"scripts": {
+					"test": "echo \"Error: no test specified\" && exit 1"
+				},
+				"author": "John doe",
+				"license": "MIT",
```
This is not necessary imo

TomChv commented on core/integration/module_typescript_test.go:
```diff
@@ -473,6 +473,128 @@ func TestModuleTypescriptOptional(t *testing.T) {
+
+	t.Run("should use package.json configuration bun", func(t *testing.T) {
+		modGen := modGen.WithNewFile("/work/dagger/package.json", dagger.ContainerWithNewFileOpts{
+			Contents: `{
+				"name": "my-module",
+				"version": "1.0.0",
+				"description": "My module",
+				"main": "index.js",
+				"scripts": {
+					"test": "echo \"Error: no test specified\" && exit 1"
+				},
+				"author": "John doe",
+				"license": "MIT",
```
not necessary too

TomChv commented on core/integration/module_typescript_test.go:
```diff
@@ -473,6 +473,128 @@ func TestModuleTypescriptOptional(t *testing.T) {
+			With(sdkSource("typescript", runtimeDetection)).
+			WithNewFile("/work/dagger/package.json", dagger.ContainerWithNewFileOpts{
+				Contents: `{
+					"name": "my-module",
+					"version": "1.0.0",
+					"description": "My module",
+					"main": "index.js",
+					"scripts": {
+						"test": "echo \"Error: no test specified\" && exit 1"
+					},
+					"author": "John doe",
+					"license": "MIT",
```
Not necessary

TomChv reviewed (changes requested):
Thanks for your amazing contribution!
I left some comments, feel free to update your PR or I can do it later if you prefer ;)

benjaminstrasser commented:
@TomChv 
I just rebased and pushed all the requested changes except the one regarding the bun caching location.
I did not find a quick way to change it or debug the behavior to be certain it works the way we intended. I reran the test with the example I gave, which gives me some confidence everything is working. Maybe you have some good ideas how to best resolve this.

TomChv commented on sdk/typescript/runtime/main.go:
```diff
@@ -146,13 +189,39 @@ func (t *TypeScriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
+	case "bun":
+		return dag.Container().
+			From("oven/bun:1.0.27").
+			WithMountedCache("~/.bun/install/cache", dag.CacheVolume("mod-bun-cache")).
```
Alright! Let's keep it that way then!

TomChv reviewed (approved):
Amazing! I approve the PR but I'll hold the merge until we complete the `v0.10.0` milestone.
So we can follow up the PR with some docs ;)

jedevc commented:
Created a v0.10.1 milestone, and added it to that, so we should be able to take this for then!

@benjaminstrasser I removed "Draft" from the PR title - hopefully that's ok, and this is ready to merge.

Could you also add a release note fragment following the instructions here? https://github.com/dagger/dagger/blob/main/CONTRIBUTING.md#3-add-release-notes-fragment. We haven't previously been changelogging zenith modules changes, but after v0.10.0 we will start doing so, so would be helpful to have a line in the changelog to show this off! :tada: Cheers for all your work on this! :rocket:

TomChv commented:
I tested my [eslint module](https://github.com/quartz-technology/daggerverse/tree/main/eslint) with all operations cached (for a fair comparison between node and bun)
```shell
# Node
time dagger call --files=src run stdout
dagger call --files=src run stdout  0.94s user 0.48s system 8% cpu 15.903 total

# Bun
➜  eslint git:(main) ✗ time dagger call --files=src run stdout
dagger call --files=src run stdout  0.63s user 0.32s system 17% cpu 5.454 total
```

Bun is really faster than node during the setup -> almost `x3` faster after caching

I wonder how fast it could be if we use it on our repository, we could probably improve our Ts CI speeds and still be on `node` but with a `bun` runtime in it.

That's really interesting!

/cc @gerhard

I think I can test that next week by replacing the default runtime we use in the template by `bun` and see if it actually increase the CI speed.

Since calling init in each of our testcase and the TS one is probably one of the slower, I'm pretty sure we'll gain 1 or 2 minutes with this super small change.

marcosnils commented:
:wave:  awesome contribution @benjaminstrasser! thx for sharing. 

I have one more fundamental question re this PR. IIUC the `DetectRuntime` function will be called each time that someone tries to run a typescript function. It seems to me that running this code here: https://github.com/dagger/dagger/pull/6730/files#diff-a8b5f7f08501c3951b9991d2450b8b0a68ac764af369185bbf7044d9654ac58cR220-R236 will add a bit of overhead to each detection cycle since it quires running a Dagger container for such purpose. After skimming the `__runtime_detection` logic (https://github.com/dagger/dagger/pull/6730/files#diff-a8a40d72e98e14628f2eb0abf1304d14cbb3dce905058aeb72736f07c47c0944) seems to me like all that code could be written in Go directly since I couldn't spot any node/bun runtime specific handling and being performed within the TypescriptSDK runtime module instead of pulling an additional image and running a container. 

Thoughts?

helderco reviewed (commented):
Thanks for this, @benjaminstrasser!

Had this review pending for a while, submitting what I have now.

helderco commented on sdk/typescript/runtime/main.go:
```diff
@@ -149,13 +199,40 @@ func (t *TypeScriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
-		From(fmt.Sprintf("node:%s", version)).
-		WithMountedCache("/root/.npm", dag.CacheVolume("mod-npm-cache-"+version)).
-		WithoutEntrypoint()
+func (t *TypeScriptSdk) DetectRuntime(ctx context.Context, modSource *ModuleSource, subPath string) (SupportedTSRuntime, error) {
+	detectedRuntime, err := dag.Container().
+		From("oven/bun:1.0.27").
```
@marcosnils noticed the same.

helderco commented on sdk/typescript/runtime/main.go:
```diff
@@ -50,17 +57,34 @@ func (t *TypeScriptSdk) ModuleRuntime(ctx context.Context, modSource *ModuleSour
 		// need to specify --no-deprecation because the default package.json has no main field which triggers a warning
 		// not useful to display to the user.
 		WithEntrypoint([]string{"tsx", "--no-deprecation", "--tsconfig", tsConfigPath, entrypointPath}), nil
+	default:
+		return nil, fmt.Errorf("unknown runtime: %v", detectedRuntime)
+	}
```
If bun is faster I'm ok making that the default instead of Node (dagger init), just not sure how the community will receive that, it takes one to know one.

I'm considering also in Python adopting a new tool as a default since it's so much faster. We need to make it as fast as possible to get to the function to get a more responsive CLI and avoid adding bottlenecks in the ecosystem.

jedevc replied:
I'm also not opposed to switching the default, but I think we should aim to land that change in a separate release from the one we first introduce it in - introducing and *potentially* breaking in the same release is something I'd like to avoid.

benjaminstrasser replied:
While I also think switching the default, just for the speed is generally a good idea; we cannot forget that most people assume node is the default runtime and while bun works with the dagger SDK not all node APIs and therefore not all npm packages work with bun yet. Most popular packages do work, but if someone uses a non-compatible npm package, the resulting error might be very confusing. (Example: jsonwebtoken uses nodes Diffie-Hellman, which is not implemented yet in bun)
This might be something we can alleviate with good documentation but still worth considering.

marcosnils replied:
I don't want to sound like the pesimist but I'd strongly advocate to **not** changing the default runtime to bun.  As @benjaminstrasser mentioned above, bun is still not fully node API compatible and there are some libraries that fail wiredly when using it

benjaminstrasser replied:
> `DetectedRuntime` already validates supported runtimes so this will never run. If defaulting to Node, it would be more like:
> 
> ```go
> if isBun {
>   // do bun
> } else {
>   // do node
> }
> ```
> 
> And later:
> 
> ```go
> switch runtime {
> case Bun:
>     ...
> case Deno:
>     ...
> default:
>     // node
> }
> ```
>

I am usually trying to be rather defensive with implicit guarantees. If the compiler cannot assure something, which is the case as DetectRuntime can return `SupportedTSRuntime("bla)"`, I am checking it at runtime. I usually prefer that the consumer of  function can only trust explicit compile-time guarantees . F.e in TS or Java allow for such compile time guarantees, but I have not found a way to do this in go. Maybe I am missing something here.

I am also completely fine with changing it however. Just a personal preference of mine.

> Another way to organize this is to implement an interface (just in Go, not exposed to API) and create different `type Node struct` and `type Bun struct` that implement that interface, returned by `DetectRuntime() Runtime` (standalone function, not a part of `TypeSriptSdk` as to not expose). This is what I plan on doing with Python.

Would you like me to look into it now or replicate your design approach for ts once you have a draft for python?

helderco replied:
> I am usually trying to be rather defensive with implicit guarantees. If the compiler cannot assure something, which is the case as DetectRuntime can return `SupportedTSRuntime("bla)"`, I am checking it at runtime. I usually prefer that the consumer of function can only trust explicit compile-time guarantees . F.e in TS or Java allow for such compile time guarantees, but I have not found a way to do this in go. Maybe I am missing something here.

It's more like, node is what we support up to now. But... if you have `"runtime": "bun"`, you can use Bun instead. If it's `"runtime": "bla"` then sure, return an error. But If it's empty, and you don't auto detect from a bun lock file, rather than looking for a node lock file, just assume node.

You can argue that adding the check for node as well can give a more useful error earlier to the user in case things get messed up somehow. So I'm not opposed to it.

> Would you like me to look into it now or replicate your design approach for ts once you have a draft for python?

A refactor isn't necessary for the PR's approval (more important is the performance concerns), unless you want to or you're inspired by the following ideas that I'm working on:
- A `Config` struct to hold configuration, instantiated in the constructor with the SDK's defaults, corresponding to the data structure in `package.json`'s `"dagger"` key;
- Using `modSource.Directory("").Entries(ctx)` to do some discoverability over which files are present (like the bun lock file) to override the defaults in the initial config, but also for checking if there's a `package.json` to begin with which replaces the WithExec that copies from the template if it doesn't exist;
- Unmarshalling `"dagger"` field from `package.json` into the `Config` instance, for explicit user settings;
- Using the final config to build the runtime Container ;

benjaminstrasser commented:
@helderco 
Thank you for the detailed review!
I will look at it in detail tomorrow. I will give some detailed input, but at  first glance everything looks like it should be implemented/adopted.

Regarding speed I already mentioned it on discord, but did not follow up out of time constraints. Right now `Runtime`gets called two times and `Codegen` four times(2 implicit calls via Runtime and 2 explicit calls) for every `dagger call`.

I like the idea of just running dagger call inside a valid project and not needing any npm i prior but I am unsure about the baked in codegen.

For me and I would love some input:
Runtime: Simply provides a valid Runtime, all dependencies installed, entry-point setup and ready to call functions inside my source.
Codegen: Initialize or modify my setup in such a way that I can use dagger call.

Right now `dagger call` kind of does both, without writing the results from codegen to disk. If i run dagger init and delete the index.ts file and then run `dagger call container-echo --string-arg yo stdout` everything works since the default `index.ts` gets implicitly generated. This does feel counter intuitive to me. I would love some input on this. Maybe I am missing something here.

TomChv commented:
> Right now dagger call kind of does both, without writing the results from codegen to disk. If i run dagger init and delete the index.ts file and then run dagger call container-echo --string-arg yo stdout everything works since the default index.ts gets implicitly generated. This does feel counter intuitive to me. I would love some input on this. Maybe I am missing something here.

If you rename `index.ts` to `foo.ts`, the file should not be replaced.
 
> For me and I would love some input:
Runtime: Simply provides a valid Runtime, all dependencies installed, entry-point setup and ready to call functions inside my source.
Codegen: Initialize or modify my setup in such a way that I can use dagger call.

Runtime actually runs codegen because you might have external dependencies that have changed, specially when you work with multiples local multiples.
By running the codegen, we ensure your module is always ready to be executed.

It's quite hard to do one without the other, because creating new functions in your module actually updates the codegen, meaning that we *must* reexecute it every time.

> Regarding speed I already mentioned it on discord, but did not follow up out of time constraints. Right now Runtimegets called two times and Codegen four times(2 implicit calls via Runtime and 2 explicit calls) for every dagger call.

That's true but thanks cache, it shouldn't take long because the codegen is actually reexecuted only 1 time, after that all operations are cached.

helderco commented:
> Regarding speed I already mentioned it on discord, but did not follow up out of time constraints. Right now `Runtime`gets called two times and `Codegen` four times(2 implicit calls via Runtime and 2 explicit calls) for every `dagger call`.
> 
> I like the idea of just running dagger call inside a valid project and not needing any npm i prior but I am unsure about the baked in codegen.
> 
> For me and I would love some input: Runtime: Simply provides a valid Runtime, all dependencies installed, entry-point setup and ready to call functions inside my source. Codegen: Initialize or modify my setup in such a way that I can use dagger call.
> 
> Right now `dagger call` kind of does both, without writing the results from codegen to disk. If i run dagger init and delete the index.ts file and then run `dagger call container-echo --string-arg yo stdout` everything works since the default `index.ts` gets implicitly generated. This does feel counter intuitive to me. I would love some input on this. Maybe I am missing something here.

If you're curious about some historical context, the initial purpose for the codegen function in the runtime module was just to re-generate a new client (`dag`) from the extended API, say by adding dependencies with `dagger use`. At this point you were required to run `dagger develop` if you made any changes and this was quite cumbersome and led to many errors because people forgot to do it or were confused by it.

The solution was to always run codegen on every call. This makes `dagger develop` optional actually, and is simply needed for local development, to allow the IDE or the linter to follow the code paths. 

At some point later, Go moved the module's code reflection from runtime into codegen time, so codegen in Go is tied to the source coude. It generates the client from the API, but also the code to handle registration, inputs and output of function calls.

But for Python and TypeScript, the latter is still being done at runtime dynamically, through the "entrypoint". So you can set up the API codegen part in the runtime module's "codegen" in a way that its cache is not invalidated by changes in the module's source. As long as there's no changes in `introspectionJson`, you'll get a cached SDK after the first time.

Finally, we also ended up using the codegen step to also create initial files for the module. The initial `package.json` and `index.ts` files. In this case, I do agree that this should only happen once during `dagger init --sdk` or `dagger develop --sdk`. I'd prefer having a separate `Init` function that would only be called initially, which would extend the "codegen". 

I'm actually experimenting on a few changes to the Python runtime module and I found myself thinking the same thing as you. I'm going to explore checking `modSource`, even with `.Entries()` if necessary, to try to detect this initial situation and only add the template files then rather than every time. Maybe that's enough instead of a separate `Init` function.

It makes more sense to me to return a clear error that code is missing, rather than removing your `index.ts` and everything running ok with the default one from the init template implicitly.

benjaminstrasser commented:
@helderco 
I just implemented all proposed changes except one were I left a comment. I would love for you to look at them and give some input if needed.
Also added a missing test case.

helderco commented on sdk/typescript/runtime/main.go:
```diff
@@ -148,14 +210,65 @@ func (t *TypeScriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
+		}
+	}
+
+	// Try to detect runtime from lock files
+	// TODO: change to ".Exists" when it's added to core
+	if _, err := source.File(path.Join(subPath, "package-lock.json")).Size(ctx); err == nil {
+		return Node, nil
+	}
+
+	// TODO: change to ".Exists" when it's added to core
+	if _, err := source.File(path.Join(subPath, "bun.lockb")).Size(ctx); err == nil {
+		return Bun, nil
 	}
```
The way I've been doing this is through `entries, _ := modSource.Directory("").Entries(ctx)`, which gives you a list of top level files in the source directory in a single request. So checking for `package-lock.json` and others just takes checking if they're in `entries`.

helderco commented on sdk/typescript/runtime/main.go:
```diff
@@ -148,14 +210,65 @@ func (t *TypeScriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
+// If none of the above is present, node will be used.
+func (t *TypeScriptSdk) DetectRuntime(ctx context.Context, modSource *ModuleSource, subPath string) (SupportedTSRuntime, error) {
+	// Try to detect runtime from package.json
+	source := modSource.ContextDirectory()
```
Instead of doing `path.Join(subPath, ...)` in lots of places:
```suggestion
	source := modSource.ContextDirectory().Directory(subPath)
```
Or just:
```suggestion
	source := modSource.Directory("")
```

helderco commented on sdk/typescript/runtime/main.go:
```diff
@@ -148,14 +210,65 @@ func (t *TypeScriptSdk) CodegenBase(ctx context.Context, modSource *ModuleSource
+		}
+
+		value := gjson.Get(json, "dagger.runtime").String()
+		if value == "" {
+			return "", fmt.Errorf("no runtime specified in package.json")
+		}
```
It looks like setting "runtime" in "package.json" is now required, even for "node", but these errors are actually ignored.

Readability could be better if in `value == ""` you go look for the lock files, and then do the switch case in scope, with a value that you know isn't empty (`switch SupportedTSRuntime(value)`).

benjaminstrasser commented:
> If you're curious about some historical context, the initial purpose for the codegen function in the runtime module was just to re-generate a new client (`dag`) from the extended API, say by adding dependencies with `dagger use`. At this point you were required to run `dagger develop` if you made any changes and this was quite cumbersome and led to many errors because people forgot to do it or were confused by it.
> 
> The solution was to always run codegen on every call. This makes `dagger develop` optional actually, and is simply needed for local development, to allow the IDE or the linter to follow the code paths.
> 
> At some point later, Go moved the module's code reflection from runtime into codegen time, so codegen in Go is tied to the source coude. It generates the client from the API, but also the code to handle registration, inputs and output of function calls.
> 
> But for Python and TypeScript, the latter is still being done at runtime dynamically, through the "entrypoint". So you can set up the API codegen part in the runtime module's "codegen" in a way that its cache is not invalidated by changes in the module's source. As long as there's no changes in `introspectionJson`, you'll get a cached SDK after the first time.
> 
> Finally, we also ended up using the codegen step to also create initial files for the module. The initial `package.json` and `index.ts` files. In this case, I do agree that this should only happen once during `dagger init --sdk` or `dagger develop --sdk`. I'd prefer having a separate `Init` function that would only be called initially, which would extend the "codegen".
> 
Thanks for the detailed description :heart: . This gave me a lot of the necessary context to understand the status quo.  I still do not understand why codegen and runtime get called twice.

> I'm actually experimenting on a few changes to the Python runtime module and I found myself thinking the same thing as you. I'm going to explore checking `modSource`, even with `.Entries()` if necessary, to try to detect this initial situation and only add the template files then rather than every time. Maybe that's enough instead of a separate `Init` function.
> 
> It makes more sense to me to return a clear error that code is missing, rather than removing your `index.ts` and everything running ok with the default one from the init template implicitly.

Yeah I totally agree. I would love to do the same or help with the typescript runtime. I feel we can optimize performance and make the runtime more intuitive.

helderco reviewed (approved):
Thanks! 👍

helderco commented:
Failing test:
```
535
=== RUN   TestModuleTypescriptRuntimeDetection/should_detect_bun.lockb
536
    module_typescript_test.go:576: 
537
        	Error Trace:	/app/core/integration/module_typescript_test.go:576
538
        	Error:      	Not equal: 
539
        	            	expected: map[string]interface {}{"runtimeDetection":map[string]interface {}{"echoRuntime":"bun"}}
540
        	            	actual  : map[string]interface {}{"runtimeDetection":map[string]interface {}{"echoRuntime":"node"}}
541
        	            	Diff:
542
        	            	--- Expected
543
        	            	+++ Actual
544
        	            	@@ -2,3 +2,3 @@
545
        	            	  (string) (len=16) "runtimeDetection": (map[string]interface {}) (len=1) {
546
        	            	-  (string) (len=11) "echoRuntime": (string) (len=3) "bun"
547
        	            	+  (string) (len=11) "echoRuntime": (string) (len=4) "node"
548
        	            	  }
549
        	Test:       	TestModuleTypescriptRuntimeDetection/should_detect_bun.lockb
550
--- FAIL: TestModuleTypescriptRuntimeDetection/should_detect_bun.lockb (24.39s)
```

helderco commented:
@benjaminstrasser, still need to fix the test.

benjaminstrasser commented:
> @benjaminstrasser, still need to fix the test.

The test you specified should be fixed now. I had some issues where the test runs were cached, even though the cache was cleared. Still some failing tests. I'll try to have a look tomorrow and ping you when everything is done.

TomChv commented:
I retriggered the test, I'm waiting for it to be green to merge the PR!
